### PR TITLE
Small supertype improvements

### PR DESCRIPTION
**Pull request by jaredvann**

### CODE REVIEW

File: phantom/agents.py [hunk 1]
```diff
@@ -161,6 +161,17 @@ def reset(self) -> None:
         Can be extended by subclasses to provide additional functionality.
         """
 
+        if hasattr(self, "Supertype"):
```
Comment (leoardon) [inline phantom/agents.py:164]: That means that here we must enforce the class to be called Supertype. Is there another way to check whether an agent has a supertype associated?

File: phantom/env.py [hunk 2]
```diff
@@ -135,27 +145,26 @@ def resolve_network(self) -> None:
...
         """
         # Reset clock
         self.current_step = 0
 
-        # Sample from supertypes and apply to agent and env type objects
-        if sample_supertypes:
-            self.sample_supertypes()
+        # Sample from supertype shared sampler objects
+        for sampler in self._samplers:
+            sampler.value = sampler.sample()
```
Comment (leoardon) [inline phantom/env.py:164]: can this assignment be done within the `sample` method of the sampler object? I find it strange to change the property of the object outside the scope of the object

File: tests/__init__.py [hunk 3]
```diff
@@ -1,11 +1,24 @@
...
 import numpy as np
 import phantom as ph
 
 
+class MockSampler(ph.utils.samplers.Sampler[float]):
+    def __init__(self, value: float) -> None:
+        self.value = value
+
+    def sample(self) -> float:
+        return self.value
```
Comment (leoardon) [inline tests/__init__.py:14]: can you change that to `self.value + 1` for example to make the test more realistic please?

File: phantom/env.py [hunk 4]
```diff
@@ -74,6 +74,11 @@ def __init__(
         self._samplers: List[Sampler] = []
 
         if env_supertype is not None:
+            if isinstance(env_supertype, dict):
+                env_supertype = self.Supertype(**env_supertype)
```
Comment (leoardon) [inline phantom/env.py:78]: here as well, we assume that the Supertype inner class is present

File: phantom/env.py [hunk 5]
```diff
@@ -86,6 +91,11 @@ def __init__(
 
         if agent_supertypes is not None:
             for agent_id, agent_supertype in agent_supertypes.items():
+                if isinstance(agent_supertype, dict):
+                    agent_supertype = self.agents[agent_id].Supertype(**agent_supertype)
```
Comment (leoardon) [inline phantom/env.py:95]: here as well.

**Review by leoardon — CHANGES_REQUESTED** [review]

File: phantom/agents.py (hunk 1 above)
Comment (jaredvann) [reply phantom/agents.py]: Fixed. Now the agent will first check for an `Agent.supertype` supertype instance, if not will then check for an `Agent.Supertype` class to generate a default instance from, if not no supertype is set.

File: phantom/env.py (hunk 2 above)
Comment (jaredvann) [reply phantom/env.py]: Fixed. Now the `.value` is a `@property` of the `Sampler` class and is set with the `.sample()` method. This is better than before as the `.value` property is now (as much as Python is) read only to users of the class

File: tests/__init__.py (hunk 3 above)
Comment (jaredvann) [reply tests/__init__.py:14]: I'm not sure how this improves the tests? The `MockSampler` is just an identity Sampler.

File: phantom/env.py (hunk 4 above)
Comment (jaredvann) [reply phantom/env.py:78]: Yes. If the supertype values are passed as a dict then we need a `Supertype` inner class to check against and build the supertype instance. The idea of this is that when evaluating a policy/environment the agent class that the supertype belongs to may not be in scope. Instead we can pass the values and ranges in as a dict.

File: tests/__init__.py (hunk 3 above)
Comment (leoardon) [reply tests/__init__.py:14]: it is to guarantee that the value is actually modified when the sampler.sample method is called, and that all the components are using the latest value as opposed to a spurious copy

**Pull request merged**